lib/src/kbasesearchengine/main/NarrativeInfoDecorator.java

```diff
@@ -96,51 +107,69 @@ public SearchObjectsOutput searchObjects(final SearchObjectsInput params, final
+     * add the removed result's guid to removedGuids list, if list is not null.
+     */
     private Map<Long, Tuple5 <String, Long, Long, String, String>> addNarrativeInfo(
             final List<ObjectData> objects,
-            final Map<Long, Tuple5 <String, Long, Long, String, String>> accessGroupNarrInfo) {
+            final Map<Long, Tuple5 <String, Long, Long, String, String>> accessGroupNarrInfo,
+            final List<String> removedGuids) {
 
         final Map<Long, Tuple5 <String, Long, Long, String, String>> retVal = new HashMap<>();
 
         if (accessGroupNarrInfo != null) {
             retVal.putAll(accessGroupNarrInfo);
         }
-        final Set<Long> wsIdsSet = new HashSet<>();
-
-        for (final ObjectData objData: objects) {
+        final Set<String> userNames = new HashSet<>();
+        final Iterator<ObjectData> iter = objects.iterator();
+        while (iter.hasNext()) {
+            final ObjectData objData = iter.next();
             final GUID guid = new GUID(objData.getGuid());
             if (WorkspaceEventHandler.STORAGE_CODE.equals(guid.getStorageCode())) {
-                wsIdsSet.add((long) guid.getAccessGroupId());
-            }
-        }
-        final Set<String> userNames = new HashSet<>();
-        for (final long workspaceId: wsIdsSet) {
-            final NarrativeInfo narrInfo = narrInfoProvider.findNarrativeInfo(workspaceId);
-            if (narrInfo != null) {
-                final Tuple5<String, Long, Long, String, String> tempNarrInfo =
-                        new Tuple5<String, Long, Long, String, String>()
-                                .withE1(narrInfo.getNarrativeName())
-                                .withE2(narrInfo.getNarrativeId())
-                                .withE3(narrInfo.getTimeLastSaved())
-                                .withE4(narrInfo.getWsOwnerUsername());
-                userNames.add(tempNarrInfo.getE4());
-                retVal.put(workspaceId, tempNarrInfo);
-            }
-            else {
-                retVal.put(workspaceId, null);
+                final long workspaceId = guid.getAccessGroupId();
+                final NarrativeInfo narrInfo = narrInfoProvider.findNarrativeInfo(workspaceId);
+
+                if (narrInfo != null) {
+                    final Tuple5<String, Long, Long, String, String> tempNarrInfo =
+                            new Tuple5<String, Long, Long, String, String>()
+                                    .withE1(narrInfo.getNarrativeName())
+                                    .withE2(narrInfo.getNarrativeId())
+                                    .withE3(narrInfo.getTimeLastSaved())
+                                    .withE4(narrInfo.getWsOwnerUsername());
+                    userNames.add(tempNarrInfo.getE4());
+                    retVal.put(workspaceId, tempNarrInfo);
+                }
+                else {
+                    iter.remove();
```

Comment:
What about refdata?

Reply:
Ref data isn't checked. This is for the case of narrative search where narrativeinfo comes back as null, the results are removed.

Reply:
Right, and for refdata I presume narrative info will be null, so all refdata will be removed from the results, right?

Reply:
Good point. How can I tell if an object is a refdata?

Reply:
are sourcetags reliable?

Reply:
The reason I ask is that this logic was already there based on the diff, and I assume we're not excluding refdata now.

Reply:
@Override
    public NarrativeInfo findNarrativeInfo(final Long wsId) {
        try {
            return cache.get(wsId);
        } catch (Exception e) {
            LoggerFactory.getLogger(getClass()).error("ERROR: Failed retrieving narrative info: Returning null:  {}",
                    e.getMessage());
            // unchecked exceptions are wrapped in UncheckedExecutionException
            return null;
        }
    }

Reply:
public NarrativeInfo findNarrativeInfo(final Long accessGroupID) {
        Tuple9 <Long, String, String, String, Long, String, String,
                String, Map<String,String>> wsInfo;

        try {
            wsInfo = wsHandler.getWorkspaceInfo(accessGroupID);
        } catch (IOException | JsonClientException e) {
            LoggerFactory.getLogger(getClass()).error("ERROR: Failed retrieving workspace info: Returning null: {}",
                    e.getMessage());
            wsInfo = null;
        }

        if (wsInfo == null) return null;

        final long timeMilli = WorkspaceEventHandler.parseDateToEpochMillis(wsInfo.getE4());

        final NarrativeInfo tempNarrInfo =
                new NarrativeInfo(null, null, timeMilli, wsInfo.getE3());

        final Map<String, String> wsInfoMeta = wsInfo.getE9();

        if (wsInfoMeta.containsKey("narrative") &&
                wsInfoMeta.containsKey("narrative_nice_name")) {
            tempNarrInfo.withNarrativeName(wsInfoMeta.get("narrative_nice_name"))
                    .withNarrativeId(Long.parseLong(wsInfoMeta.get("narrative")));
        }
        return tempNarrInfo;
    }

Reply:
ugh can't get the formatting right

Reply:
looks like if only returns nulls if workspace throughs error. For a refdata workspace, would the rest of code error?

Reply:
This also seems odd, since it means the results returned from the search API may be different over time.

But in any case, non-narrative results, based on this code, is not removed.

> For a refdata workspace, would the rest of code error?

Do you see any reason it would?

Reply:
what do you mean different over time? 

idk. would fields from the wsInfo such as E9 (metadata) ever be null?

Reply:
Wait a sec, this makes no sense. If there's a network blip and the workspace can't be contacted then no data will be returned.

Reply:
You mean wsInfo would be empty? wsHandler wouldn't through an error?

Reply:
if you look at the code above, if *any* error occurs contacting the ws, narrinfo is set to null, which means the data is removed from the results returned by the search API.

Reply:
isn't that what we want? If there is a network blip and we won't get any data, the we can't show the data on the front end anyway.

Reply:
I guess we can do retries. but if we are going to do that, there are so many other areas that could use this...

Reply:
> isn't that what we want? If there is a network blip and we won't get any data, the we can't show the data on the front end anyway.

No, the difference here is that the user will think there's no results for their search rather than there was an error with their search. If there's a blip we should retry or throw an error, not tell the user there aren't any results.

Reply:
what would no results look like?

Reply:
I think this fix is going in another PR

Reply:
Should be in this pr. Right now for this pr we are assuming that a null result from a workspace means it has been deleted. The next Pr should be handling the different errors (deletion vs non deletion). And if there needs a different check for handling deleted workspaces, it should be changed later.

Reply:
> The next Pr should be handling the different errors (deletion vs non deletion)

that's the fix I meant. The original issue isn't an issue if we assume that the narrative info provider will be fixed such that it only returns null if the narrative is deleted.

Reply:
Although we should probably document - very clearly - what's going on here and when the narrative info provider is supposed to return null and when it isn't.

Reply:
But the original issue is that we still want deleted objects to be removed from in the results?

Reply:
The original issue from the top comment in this thread is

> What about refdata?

Reply:
refdata would still be included? it should still have workspace info correct?

Reply:
I don't really understand what you're asking here.

My understanding is that previous to this PR, refdata would not be excluded from the returned results because the narrativeInfoProvider doesn't actually provide just narrative info. It provides info about the workspace, and also provides additional info about the narrative if the workspace is a narrative. So narrInfo is only null if the workspace isn't accessible, not if the workspace is not a narrative, which is unintuitive, which is why I thought refdata would be removed from the result. 

Since my intuition was wrong, I don't think the original concern about refdata in the original comment was an issue, and nothing in this PR changes that.

What we should do is 

> document - very clearly - what's going on here and when the narrative info provider is supposed to return null and when it isn't.

Reply:
okay will add comments explaining this